### ISSUE TITLE
Complex->Real cast is a warning, calling `real` or `imag` on non-comlex tensors is an Error.

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#51210]
As per title.

I would expect exactly the opposite behavior, that is:
Casting complex tensors to real ones is an error. And if this cast is necessary, then it makes senses to call it on either real or imaginary part.
I see no issues on calling `real` or `imag` on real tensors as these operations are well-defined.

Current behavior:
```python
In [1]: import torch
x
In [2]: x = torch.rand(3, 3, dtype=torch.cdouble).long()
/home/nik/.conda/envs/pytorch-cuda-dev/bin/ipython:1: UserWarning: Casting complex values to real discards the imaginary part (Triggered internally at  ../aten/src/ATen/native/Copy.cpp:219.)
  #!/home/nik/.conda/envs/pytorch-cuda-dev/bin/python

In [3]: x = torch.rand(3, 3)

In [4]: x.imag
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-4-f2d086c634a9> in <module>
----> 1 x.imag

RuntimeError: imag is not implemented for tensors with non-complex dtypes.

In [5]: x.real
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-afe486272bd1> in <module>
----> 1 x.real

RuntimeError: real is not implemented for tensors with non-complex dtypes.

```

**Comment by anjali411** [comment]
@nikitaved fair point! we disabled `real` and `imag` for real tensors because while `real` can be a view for real tensors as it is for complex, `imag` can not. And so, we decided to disable both ops for real tensors. We are acting conservatively because it's easier to enable ops than disabling them (due to bc).

**Comment by nikitaved** [comment]
I see, but still casting to reals is allowed.
Real tensors on `imag` could return an empty Tensor...

**Comment by gchanan** [comment]
> @nikitaved fair point! we disabled real and imag for real tensors because while real can be a view for real tensors as it is for complex, imag can not. And so, we decided to disable both ops for real tensors. We are acting conservatively because it's easier to enable ops than disabling them (due to bc).

I don't get the view point for complex tensors.  Presumably you want it to be read-only and it might be most fruitful to implement it as a read-only-view, but it's not really a view from the perspective of the real tensor AFAICT.
